codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -230,19 +270,48 @@ private int connectionWindowSize() {
         return connectionState().windowSize();
     }
 
+    private void resetConnectionWindowHandicap() {
+        DefaultConnectionState connectionState = connectionState();
+        if (ctx == null) {
+            connectionState.windowSizeHandicap(connectionState.windowSizeWithoutHandicap());
+        } else {
+            int channelWritableBytes = (int) Math.min(Integer.MAX_VALUE, ctx.channel().bytesBeforeUnWritable());
+            // The current allocation algorithm values "fairness" and doesn't give any consideration to "goodput". It
+            // is possible that 1 byte will be allocated to many streams. In an effort to try to make "goodput"
+            // reasonable with the current allocation algorithm we have this "cheap" check up front to ensure there is
+            // an "adequate" amount of connection window before allocation is attempted. This is not foolproof as if the
+            // number of streams is >= this minimal number then we will have the issue, but the idea is to narrow the
+            // circumstances in which this can happen without rewriting the allocation algorithm.
+            int minimalUsableChannelBytes =
+                    Math.max(ctx.channel().config().getWriteBufferHighWaterMark(), 1 << 15) >>> 1;
+            connectionState.windowSizeHandicap(channelWritableBytes < minimalUsableChannelBytes ?
+                    connectionState.windowSizeWithoutHandicap() :
+                    connectionState.windowSizeWithoutHandicap() -
+                    Math.min(connectionState.windowSizeWithoutHandicap(), channelWritableBytes));
+        }
+    }
+
     /**
      * Writes as many pending bytes as possible, according to stream priority.
      */
     private void writePendingBytes() throws Http2Exception {
-        Http2Stream connectionStream = connection.connectionStream();
-        int connectionWindowSize = state(connectionStream).windowSize();
+        assert !isWriting;
+        isWriting = true;
```

Comment:
@nmittler - I tried to clarify when we ran out of time the other day :)  I don't think we have to loop (and continuously check `isWritable()`) if we know how many bytes we can write until writability changes. Instead we can check 1 time upfront, and since we are single threaded we should be able to exhaust this amount in our allocation loop.

Either way we need to be able to protect against re-entry into the allocation/write loop.